mypy/types.py

```diff
@@ -868,7 +959,7 @@ def __init__(self,
         assert len(arg_types) == len(arg_kinds) == len(arg_names)
         if variables is None:
             variables = []
-        self.arg_types = arg_types
+        self.arg_types = list(arg_types)
```

Comment:
What about making sure that these are expanded enough so that they are proper types (but may contain type alias references)? Not sure if this would be worth it, but it might let us avoid a bunch of `get_proper_type` calls elsewhere.

Reply:
It looks like this is actually not possible. Actually even what I did for unions is problematic, for example if we have (not very meaningful, since technically `A = int`) `A = Union[int, A]`. To represent this we must have somewhere a union that contains an item that is `A`. I will try to fix this now.

Reply:
Actually these kind of "trivial" recursive unions are the most problematic, maybe we should actually ban them. Another solution would be to actually remove aliases being defined from immediate union, but this can silently hide an error, so I don't like this.

Reply:
After some more thinking I think I convinced myself that aliases like `A = Union[int, A]` are really bad. Essentially they are not much different from `A = A`, for example all situations like this:
```python
x: A
x.to_bytes()
```
cause infinite recursion. Also, it is very hard to imagine any practical use for such aliases. Finally, they can be easily detected during semantic analysis and banned.

If we ban them, then we can keep `UnionType.items` and `TypeType.item` proper. I however didn't find any simple consistent representation schema where we could also make `TupleType.items` and `CallableType.arg_types` (and `.ret_type`) proper, mostly because of things like `A = Union[int, Callable[[], A]]`.

@JukkaL If you don't have objections, I would merge this to get things moving (sorry for pinging during vacation :-)).